File: core/core.net/source/core.net/IO/BinaryReader/BinaryReader.cs
//
// Copyright (c) 2015. SeongJun Park.
//
// Distributed under the MIT License.
//

using System;
using System.Text;
using System.Runtime.InteropServices;
using System.Collections;
using System.Reflection;
using System.IO;

namespace Useless.IO
{
	public class BinaryReader : IDisposable
	{
		class StreamProxy
		{
			private IStream m_uselessStream = null;
			private Stream m_stream = null;

			public StreamProxy( IStream stream )
			{
				m_uselessStream = stream;
			}

			public StreamProxy( Stream stream )
			{
				m_stream = stream;
			}

			public void Dispose()
			{
				if( m_uselessStream != null )
				{
					m_uselessStream.Dispose();
				}
				else if( m_stream !=  null )
				{
					m_stream.Dispose();
				}
			}

			public object GetStream()
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream;
				}
				else
				{
					return m_stream;
				}
			}

			public int Seek( int offset, SeekDir way )
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream.Seek( offset, way );
				}
				else
				{
					return ( int )m_stream.Seek( offset, ( SeekOrigin )way );
				}
			}

			public int GetPosition()
			{
				if( m_uselessStream != null )
				{

					return m_uselessStream.Position;
				}
				else
				{
					return ( int )m_stream.Position;
				}
			}

			public int Read( byte[] buffer, int count )
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream.Read( buffer, count );
				}
				else
				{
					return m_stream.Read( buffer, 0, count );
				}
			}
		}

		private static byte[] m_internalBuffer = new byte[ 8 ];

		private StreamProxy m_stream = null;
		private bool m_mustBeDispose = false;

		public BinaryReader()
		{
			
        }

		public BinaryReader( IStream stream )
		{
			if( stream == null || !stream.CanRead )
			{
				throw new ArgumentException( "stream" );
			}

			m_stream = new StreamProxy( stream );
		}

		public BinaryReader( Stream stream )
		{
			if( stream == null || !stream.CanRead )
			{
				throw new ArgumentException( "stream" );
			}

			m_stream = new StreamProxy( stream );
		}

		public BinaryReader( byte[] address, int size )
		{
			m_stream = new StreamProxy( new MemoryStream( address, size ) );
			m_mustBeDispose = true;
        }

		public BinaryReader( String path, System.IO.FileMode openmode )
		{
			m_stream = new StreamProxy( new FileStream( path, openmode ) );
			m_mustBeDispose = true;
		}

		public void Dispose()
		{
			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = null;
		}

		public void SetStream( IStream stream )
		{
			if( stream == null || !stream.CanRead )
			{
				throw new ArgumentException( "stream" );
			}

			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = new StreamProxy( stream );
			m_mustBeDispose = false;
        }

		public void SetStream( Stream stream )
		{
			if( stream == null || !stream.CanRead )
			{
				throw new ArgumentException( "stream" );
			}

			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = new StreamProxy( stream );
			m_mustBeDispose = false;
		}

		public object GetStream()
		{
			return m_stream.GetStream();
		}

		public int Seek( int offset, SeekDir way )
		{
			return m_stream.Seek( offset, way );
		}

		public int GetPosition()
		{
			return m_stream.GetPosition();
        }

		public void Read( byte[] buffer, int count )
		{
			if( m_stream == null )
			{
				throw new ObjectDisposedException( "stream" );
			}

			m_stream.Read( buffer, count );
		}
		
		public void Read( Type type, ref object val )
		{
			if( TypeTraits.IsArithmetic.Invoke( type ) || TypeTraits.IsChar.Invoke( type ) )
			{
				Read( m_internalBuffer, Marshal.SizeOf( type ) );
				val = Utility.BitConverter.To( type, m_internalBuffer );
			}
			else if( TypeTraits.IsString.Invoke( type ) )
			{
				uint countUnsigned = ReadU32();
				if( countUnsigned > int.MaxValue )
				{
					throw new NotSupportedException( "count" );
				}

				int count = ( int )countUnsigned;
				byte[] buffer = new byte[ count ];
				Read( buffer, count );
				val = Encoding.UTF8.GetString( buffer );
			}
			else if( type.IsArray )
			{
				long count = ( long )ReadU32();
				Type element_type = type.GetElementType();

				Array arrayVal = ( Array )val;
				for( long i = 0; i < count; ++i )
				{
					object temp = null;
					Read( element_type, ref temp );
					arrayVal.SetValue( temp, i );
				}
			}
			else if( val is IDictionary )
			{
				uint countUnsigned = ReadU32();
				if( countUnsigned > int.MaxValue )
				{
					throw new NotSupportedException( "count" );
				}

				Type key_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];
				Type value_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 1 ];
				IDictionary dictionaryVal = ( IDictionary )val;
				int count = ( int )countUnsigned;
                for( int i = 0; i < count; ++i )
				{
					object key_temp = null;
					Read( key_type, ref key_temp );

					object value_temp = null;
					Read( value_type, ref value_temp );
					dictionaryVal.Add( key_temp, value_temp );
                }
            }
			else if( val is IList )
			{
				uint countUnsigned = ReadU32();
				if( countUnsigned > int.MaxValue )
				{
					throw new NotSupportedException( "count" );
				}

				Type element_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];
				IList listVal = ( IList )val;
				int count = ( int )countUnsigned;
				for( int i = 0; i < count; ++i )
				{
					object temp = null;
					Read( element_type, ref temp );
					listVal.Add( temp );
				}
			}
			else if( val is IEnumerable || val is ICollection )
			{
				uint countUnsigned = ReadU32();
				if( countUnsigned > int.MaxValue )
				{
					throw new NotSupportedException( "count" );
				}

				Type element_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];
				int count = ( int )countUnsigned;

				MethodInfo method = type.GetMethod( "Add", new Type[] { element_type } );
				if( method == null )
				{
					method = type.GetMethod( "AddLast", new Type[] { element_type } );
                }
				if( method == null )
				{
					method = type.GetMethod( "Enqueue", new Type[] { element_type } );
				}
				if( method == null )
				{
					method = type.GetMethod( "Push", new Type[] { element_type } );
				}

				if( method != null )
				{
					for( int i = 0; i < count; ++i )
					{
						object temp = null;
						Read( element_type, ref temp );
						method.Invoke( val, new object[] { temp } );
					}
				}
				else
				{
					throw new NotSupportedException();
				}
            }
			
			else
			{
				ISerializable serializable = val as ISerializable;
				if( serializable == null )
				{
					throw new ObjectSerializeException();
				}

				BinaryReader refThis = this;
				serializable.Deserialize( ref refThis );
			}
		}

		public void Read<T>( ref T val )
		{
			object temp = val;
			Read( typeof( T ), ref temp );
			val = ( T )temp;
		}
		
		public T Read<T>()
		{
			T val = default( T );
			Read<T>( ref val );
			return val;
		}

		public bool ReadBoolean()
		{
			return Read<bool>();
		}

		public sbyte ReadS8()
		{
			return Read<sbyte>();
		}

		public byte ReadU8()
		{
			return Read<byte>();
		}
		
		public short ReadS16()
		{
			return Read<short>();
		}

		public ushort ReadU16()
		{
			return Read<ushort>();
		}
		
		public int ReadS32()
		{
			return Read<int>();
		}

		public uint ReadU32()
		{
			return Read<uint>();
		}

		public long ReadS64()
		{
			return Read<long>();
		}

		public ulong ReadU64()
		{
			return Read<ulong>();
		}

		public float ReadF32()
		{
			return Read<float>();
		}

		public double ReadF64()
		{
			return Read<double>();
		}

		public String ReadString()
		{
			return Read<String>();
		}
	}
}

File: core/core.java/source/core.java/io/src/useless/io/BinaryReader.java
/*
 * Copyright (c) 2015. SeongJun Park.
 *
 * Distributed under the MIT License.
 */

package useless.io;

import useless.io.exception.*;
import useless.type_traits.*;
import useless.utility.*;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.Array;

public class BinaryReader
{
	private StreamBase m_stream = null;
	private static byte[] m_internalBuffer = new byte[ 8 ];

	public BinaryReader()
	{

	}

	public BinaryReader( StreamBase stream )
	{
		open( stream );
	}

	public BinaryReader( byte[] address, int size )
	{
		open( address, size );
	}

	public void open( StreamBase stream )
	{
		if( stream == null || !stream.canRead() )
		{
			throw new IllegalArgumentException( "stream" );
		}

		m_stream = stream;
	}

	public void open( byte[] address, int size )
	{
		m_stream = new MemoryStream( address, size );
	}

	public StreamBase getStream()
	{
		return m_stream;
	}

	public void read( byte[] buffer, int count )
	{
		if( m_stream == null  )
		{
			throw new NullPointerException( "stream" );
		}

		int read = m_stream.read( buffer, count );
		if( read != count )
		{
			throw new ReadFailedException();
		}
	}

	@SuppressWarnings( "unchecked" )
	public <T> T read( Class<T> type ) throws UnsupportedEncodingException, InstantiationException, IllegalAccessException, ObjectSerializeException
	{
		if( IsArithmetic.invoke( type ) || IsChar.invoke( type ) )
		{
			read( m_internalBuffer, SizeOf.invoke( type ) );
			return BitConverter.to( type, m_internalBuffer );
		}
		else if( IsString.invoke( type ) )
		{
			return ( T )readString();
	 	}
		else if( type.isArray() )
		{
			return ( T )readArray( type );
		}
		else
		{
			T val = type.newInstance();
			if( val instanceof Serializable )
			{
				( ( Serializable )val ).deserialize( this );
				return val;
			}
			else
			{
				throw new ObjectSerializeException();
			}
		}
	}

	public boolean readBoolean()
	{
		read( m_internalBuffer, 1 );
		return BitConverter.toBoolean( m_internalBuffer );
	}

	public byte readS8()
	{
		read( m_internalBuffer, 1 );
		return BitConverter.toS8( m_internalBuffer );
	}

	public short readU8()
	{
		read( m_internalBuffer, 1 );
		return BitConverter.toU8( m_internalBuffer );
	}

	public short readS16()
	{
		read( m_internalBuffer, 2 );
		return BitConverter.toS16( m_internalBuffer );
	}

	public int readU16()
	{

		read( m_internalBuffer, 2 );
		return BitConverter.toU16( m_internalBuffer );
	}

	public int readS32()
	{
		read( m_internalBuffer, 4 );
		return BitConverter.toS32( m_internalBuffer );
	}

	public long readU32()
	{
		read( m_internalBuffer, 4 );
		return BitConverter.toU32( m_internalBuffer );
	}

	public long readS64()
	{
		read( m_internalBuffer, 8 );
		return BitConverter.toS64( m_internalBuffer );
	}

	public float readF32()
	{
		read( m_internalBuffer, 4 );
		return BitConverter.toF32( m_internalBuffer );
	}

	public double readF64()
	{
		read( m_internalBuffer, 8 );
		return BitConverter.toF64( m_internalBuffer );
	}

	public String readString() throws UnsupportedEncodingException
	{
		long countUnsigned = readU32();
		if( countUnsigned > Integer.MAX_VALUE )
		{
			throw new UnsupportedOperationException( "count" );
		}

		int count = ( int )countUnsigned;
		byte[] buffer = new byte[ count ];
		read( buffer, count );
		return new String( buffer, "UTF-8" );
	}

	public Object readArray( Class<?> type ) throws UnsupportedEncodingException, InstantiationException, IllegalAccessException, ObjectSerializeException
	{
		Class<?> element_type = type.getComponentType();
		if( IsArithmetic.invoke( element_type ) )
		{
			if( element_type == Boolean.class || element_type == boolean.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				boolean[] ba = new boolean[ count ];

				for( int i = 0; i < count; ++i )
				{
					ba[ i ] = readBoolean();
				}

				return ba;
			}
			else if( element_type == Byte.class || element_type == byte.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				byte[] ba = new byte[ count ];

				for( int i = 0; i < count; ++i )
				{
					ba[ i ] = readS8();
				}

				return ba;
			}
			else if( element_type == Short.class || element_type == short.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				short[] sa = new short[ count ];

				for( int i = 0; i < count; ++i )
				{
					sa[ i ] = readS16();
				}

				return sa;
			}
			else if( element_type == Integer.class || element_type == int.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				int[] ia = new int[ count ];

				for( int i = 0; i < count; ++i )
				{
					ia[ i ] = readS32();
				}

				return ia;
			}
			else if( element_type == Long.class || element_type == long.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				long[] la = new long[ count ];

				for( int i = 0; i < count; ++i )
				{
					la[ i ] = readS64();
				}

				return la;
			}
			else if( element_type == Float.class || element_type == float.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				float[] fa = new float[ count ];

				for( int i = 0; i < count; ++i )
				{
					fa[ i ] = readF32();
				}

				return fa;
			}
			else if( element_type == Double.class || element_type == double.class )
			{
				long countUnsigned = readU32();
				if( countUnsigned > Integer.MAX_VALUE )
				{
					throw new UnsupportedOperationException( "count" );
				}

				int count = ( int )countUnsigned;
				double[] da = new double[ count ];

				for( int i = 0; i < count; ++i )
				{
					da[ i ] = readF64();
				}

				return da;
			}
			else
			{
				throw new UnsupportedOperationException();
			}
		}
		else if( IsChar.invoke( element_type ) )
		{
			long countUnsigned = readU32();
			if( countUnsigned > Integer.MAX_VALUE )
			{
				throw new UnsupportedOperationException( "count" );
			}

			int count = ( int )countUnsigned;
			char[] ca = new char[ count ];

			for( int i = 0; i < count; ++i )
			{
				read( m_internalBuffer, 2 );
				ca[ i ] = BitConverter.toChar( m_internalBuffer );
			}

			return ca;
		}
		else
		{
			long countUnsigned = readU32();
			if( countUnsigned > Integer.MAX_VALUE )
			{
				throw new UnsupportedOperationException( "count" );
			}

			int count = ( int )countUnsigned;
			Object arrayVal = Array.newInstance( element_type, count );

			for( int i = 0; i < count; ++i )
			{
				Array.set( arrayVal, i, read( element_type ) );
			}

			return arrayVal;
		}
	}
}
